### ISSUE TITLE
Formatter removes semicolon after variable initializer at class level (C#)

### ISSUE THREAD

**Issue by mihaipopescu** [uncrustify/uncrustify#1333]
```cs
namespace Unity
{
    public class Class
    {
        // doesn't work because ; gets removed but
        public static readonly Class A = new Class() { name = "A", id = 1 };
        // works and ; doesn't get removed
        public static readonly Class B = new Class { name = "B", id = 2 };
    }
}
```

Expected versus output diff:
---
```diff
diff --git a/Users/mihai/Development/external/uncrustify/tests/output/staging/60016-UNI-11662.cs b/Users/mihai/Development/external/uncrustify/tests/results/staging/60016-UNI-11662.cs
index 9a49fdb..9288ee8 100644
--- a/Users/mihai/Development/external/uncrustify/tests/output/staging/60016-UNI-11662.cs
+++ b/Users/mihai/Development/external/uncrustify/tests/results/staging/60016-UNI-11662.cs
@@ -1,10 +1,10 @@
 namespace Unity
 {
     public class Class
     {
         // doesn't work because ; gets removed but
-        public static readonly Class A = new Class() { name = "A", id = 1 };
+        public static readonly Class A = new Class() { name = "A", id = 1 }
         // works and ; doesn't get removed
         public static readonly Class B = new Class { name = "B", id = 2 };
     }
 }
```

Debug support files:
---
[60016-UNI-11662.cs](https://gist.githubusercontent.com/mihaipopescu/7bb6852d950b511a706099e1f677e62f/raw/9a49fdbac1d8f0a25a778f0e8805e4946f7dc12e/60016-UNI-11662.cs)
[60016-UNI-11662.cs.log](https://gist.githubusercontent.com/mihaipopescu/7bb6852d950b511a706099e1f677e62f/raw/b79159dfe0b8ae39660e5dd5a6d95011e47d2a93/60016-UNI-11662.cs.log)
[60016-UNI-11662.cs.unc](https://gist.githubusercontent.com/mihaipopescu/7bb6852d950b511a706099e1f677e62f/raw/5934c8be6a31a078833b56ad44bb8418712fac6e/60016-UNI-11662.cs.unc)
[Uncrustify.CSharp.cfg](https://gist.githubusercontent.com/mihaipopescu/7bb6852d950b511a706099e1f677e62f/raw/1d064c1fabb76ce328dd015ed40023207de20ca1/Uncrustify.CSharp.cfg)
[UNI-11662.cs](https://gist.githubusercontent.com/mihaipopescu/7bb6852d950b511a706099e1f677e62f/raw/9a49fdbac1d8f0a25a778f0e8805e4946f7dc12e/UNI-11662.cs)

Disclaimer:
---
Note: This is a generated issue from our internal issue tracker.
Expect a dry description which could be insufficient to properly understand the issue.
On the other hand, the support files should be sufficient to easily reproduce and debug the issue,
but please comment if it's still unclear and I will get back with clarifications.
See main discussion thread here #1223.

**Comment by mihaipopescu** [comment]
This is high-priority on our side as it's actually breaking the code.

**Comment by mihaipopescu** [comment]
As probably you already may have noticed the responsible option is `mod_remove_extra_semicolon`.

**Comment by guy-maurel** [comment]
I can reproduce...

**Comment by guy-maurel** [comment]
The remove is caused by the option: `mod_remove_extra_semicolon  = true`.
Using the command option -p, one can see:
```
 FUNC_CLASS_DEF[                        NONE][ 46/ 46/ 51/  1]    Class   
       FPAREN_OPEN[  FUNC_CLASS_DEF][ 51/ 51/ 52/  0]             (    
     FPAREN_CLOSE[  FUNC_CLASS_DEF][ 52/ 52/ 53/  0]               )
```
I don't have any compiler for C#. Sorry. I don't have experience about C#.
Is the classification as `FUNC_CLASS_DEF` correct?
Could you modify the source to a cpp, and complete it to be compilable?

**Comment by mihaipopescu** [comment]
Definitely that's not a class function definition but an initialized static member.

It's kind of hard to try to reproduce the same behavior in C++ but we did saw it in something similar in our C++/CLI code as well but it might be a totally different issue:

```cpp
IIterable<int>^ Class::Foo(int first, int second)
{
	auto array = ref new Platform::Array<int>
	{
		first,
		second
	};
	return ref new Iterable<int>(array);
}
```

The semicolon after `return` statement was removed.

**Comment by guy-maurel** [comment]
Could you push a more complete source, which could be compiled?

**Comment by mihaipopescu** [comment]
Why do you need it to be compilable ? The snippets are syntactically correct and uncrustify should be able to format them just like that.

**Comment by guy-maurel** [comment]
I don't need!
I wish it!
Using the clang compiler I can see what it does.

**Comment by mihaipopescu** [comment]
I said.. it's very hard to try to come up with a C++ equivalent as this is a C# bug (the tag you put is wrong).

You can see the correct tokenization if you look at the second line where the semicolon is not removed.

**Comment by DieHertz** [comment]
For C++ `return {};` transforms to `return {}`
Using a binary distribution from Ubuntu 16.04.

**Comment by gmaurel** [comment]
Fixed at e8dd4d0d200443a021b8d99dbc384c93b977789f, Dec 31 12:38:29 2017
